tox.ini

```
@@ -3,66 +3,110 @@ skip_missing_interpreters = True
     pep8,
     codestyle,
     docstyle,
     mypy,
     packaging,
-    vulture
+    vulture,
+    # For running pytest locally
+    test-{x11,wayland,both}
 
-[testenv]
-# This is required in order to get UTF-8 output inside of the subprocesses
-# that our tests use.
-setenv = LC_CTYPE = en_US.UTF-8
-# Pass Display down to have it for the tests available
-passenv = DISPLAY,WAYLAND_DISPLAY,LDFLAGS,CFLAGS
-allowlist_externals = 
-    */ffibuild
-    convert
-# xcffib has to be installed before cairocffi
+# Set up some variables that can be used multiple times
+[base]
 deps =
-    pytest >= 6.2.1
-    coverage
     setuptools >= 40.5.0
-    bowler
     xkbcommon >= 0.3
-    pywayland == 0.4.15
     dbus-next
-    PyGObject
     wheel
     cffi
     xcffib >= 1.4.0
     cairocffi >= 1.6.0
-# pywayland has to be installed before pywlroots
+    # We use "!x11" here so pywayland is installed both in CI wayland environment
+    # But also when users run locally (as both backends are run in that scenario)
+    !x11: pywayland==0.4.15
+testdeps = 
+    pytest >= 6.2.1
+    {wayland,x11}: coverage
+    bowler
+    PyGObject   
 commands =
-    # cffi's binary wheel is incompatible with some libffi libraries so we force build here
-    # See: https://github.com/Kozea/cairocffi/issues/202
-    !pypy3: pip install --force-reinstall --no-binary :all: cffi
-    # However, rebuilding on pypy doesn't work so we'll pin the version instead
-    # See: https://github.com/tych0/xcffib/issues/134
-    pypy3: pip install cffi==1.15.0
-    ; pip install xcffib>=1.4.0 wheel
-    # The 1.5.0 release of cairocffi cannot be built in a clean folder as it will fail to
-    # to build with xcb suppport. https://github.com/Kozea/cairocffi/issues/212
-    pip install pywlroots==0.16.4
+    !x11: pip install pywlroots==0.16.4
     pip install .
     {toxinidir}/scripts/ffibuild
+
+# These are the environments that should be triggered by Github Actions
+[testenv:py{py3,39,310,311}-{wayland,x11}]
+# This is required in order to get UTF-8 output inside of the subprocesses
+# that our tests use.
+setenv = LC_CTYPE = en_US.UTF-8
+# Pass Display down to have it for the tests available
+passenv = DISPLAY,WAYLAND_DISPLAY,LDFLAGS,CFLAGS,BACKEND
+allowlist_externals = 
+    */ffibuild
+    convert
+deps =
+    {[base]deps}
+    {[base]testdeps}
+# pywayland has to be installed before pywlroots
+commands =
+    {[base]commands}
+
+    # GithubActions will trigger one of these three lines as the
+    # backend environment is set.
     # pypy3 is very slow when running coverage reports so we skip it
-    pypy3: python3 -m pytest --backend=x11 --backend=wayland {posargs}
-    py39: coverage run -m pytest --backend=x11 --backend=wayland {posargs}
-    py3{10,11}: coverage run -m pytest --backend=x11 --backend=wayland {posargs}
+    pypy3-{wayland,x11}: python3 -m pytest --backend={env:BACKEND} {posargs}
+    py39-{wayland,x11}: coverage run -m pytest --backend={env:BACKEND} {posargs}
```

Would something along these lines allow for running just the tox command without the env var?:
```suggestion
    pypy3-x11: python3 -m pytest --backend=x11 {posargs}
    pypy3-wayland: python3 -m pytest --backend=wayland {posargs}
    py39-x11: coverage run -m pytest --backend=x11 {posargs}
    py39-wayland: coverage run -m pytest --backend=wayland {posargs}
```

---

You do still need the env var in ci.yml but your change above is definitely neater.

---

Does this remove the need for users to set the env var locally?

---

I don't think they ever did. The env var is set by CI. What have I missed?

---

From the original PR message:

> One issue I have with this PR is that it makes it a bit harder for users to runtox locally. This is because, to save duplicating lines, I set the --backend option by using an environmental variable (which is set in Github Actions). To run tests, users would need to run BACKEND=wayland tox -e py311.

---

Ah, ok! Let me check but I don't think that's needed. I created some helper environment names and referred to these in the docs (e.g. `tox -e test-both`).

However, thinking about this some more, the split leads to some inefficiency if these are run in sequence as both the Wayland and x11 tests will run the same backend-agnostic tests.

---

Presumably someone would always run the both env, and not the x11 and then wayland ones separately? At least hopefully - in the former case the non-backend related tests would only be run once fortunately